apollo-federation/src/supergraph/mod.rs

```diff
@@ -1696,6 +1698,140 @@ lazy_static! {
     };
 }
 
+trait CompositeType {
+    fn directives(&self) -> &DirectiveList;
+    fn insert_directive(
+        &self,
+        schema: &mut FederationSchema,
+        pos: &CompositeTypeDefinitionPosition,
+        directive: Component<Directive>,
+    ) -> Result<(), FederationError>;
+}
+
+impl CompositeType for UnionType {
+    fn directives(&self) -> &DirectiveList {
+        &self.directives
+    }
+    fn insert_directive(
+        &self,
+        schema: &mut FederationSchema,
+        pos: &CompositeTypeDefinitionPosition,
+        directive: Component<Directive>,
+    ) -> Result<(), FederationError> {
+        match pos {
+            CompositeTypeDefinitionPosition::Union(pos) => {
+                pos.insert_directive(schema, directive.into())?;
+                return Ok(());
+            }
+            _ => Err(internal_error!("an error")),
+        }
+    }
```

Comment:
Is there error necessary here? You don't actually use `&self` in the method. Honestly, this looks like something that could be a method on `CompositeTypeDefinitionPosition`.